File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {boolean}
 */
var containsNearbyDuplicate = function (nums, k) {
  const intervalNums = new Set()
  for (let i = 0; i < k && i < nums.length; i++) {
    if (intervalNums.has(nums[i])) {
      return true
    }
    intervalNums.add(nums[i])
  }

  for (let i = k; i < nums.length; i++) {
    if (intervalNums.has(nums[i])) {
      return true
    }
    intervalNums.add(nums[i])
    intervalNums.delete(nums[i - k])
  }

  return false
}

/*
create a set to track nums

add the first k nums to the set
  if a collision happens -> return true
  
iterate over nums[k] to the end
  remove num[i - k]
  add the num at current index i
  collision? -> true

return false
*/

File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.go
func containsNearbyDuplicate(nums []int, k int) bool {
	seen := map[int]struct{}{}

	for i, num := range nums {
		if _, exists := seen[num]; exists {
			return true
		}
		seen[num] = struct{}{}
		if i >= k {
			delete(seen, nums[i-k])
		}
	}

	return false
}
